src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleReference.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// Provides functionality to serialize objects or value types to JSON and
+    /// deserialize JSON into objects or value types.
+    /// </summary>
+    public static partial class JsonSerializer
+    {
+        private static void HandleMetadataProperty(JsonSerializerOptions options, ref Utf8JsonReader reader, ref ReadStack state)
+        {
+            if (state.Current.Drain) //Verify if this is the right condition. Why should I use this over state.Current.SkipProperty?
+            {
+                return;
+            }
+
+            MetadataPropertyName meta = GetMetadataPropertyName(reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan);
+            MetadataPropertyName previous = state.Current.LastMetaProperty;
+
+            if (meta == MetadataPropertyName.Unknown)
+            {
+                if (previous == MetadataPropertyName.Ref)
+                {
+                    throw new JsonException("Properties other than $ref are not allowed in reference objects.");
+                }
+            }
+
+            else if (meta == MetadataPropertyName.Id)
+            {
+                //TODO: Validate that multiple ids are not found in the same object, however, this is not validated by Json.Net.
+            }
+
+            else if (meta == MetadataPropertyName.Values)
+            {
+                if (previous != MetadataPropertyName.Id)
+                {
+                    throw new JsonException("Cannot have a preserved enumerable without specifying one id");
+                }
+                // remove misconcepted object.
+                InitTask task = state.RemoveInitTask(state.PendingTasksCount - 1);
+                state.Current.EnumerableMetadataId = task.MetadataId;
+            }
+
+            else if (meta == MetadataPropertyName.Ref)
+            {
+                // since pending objects are initialized when a non-metadata property is found, having zero pending tasks means that $ref is not the first property in the object.
+                if (state.PendingTasksCount == 0)
+                {
+                    throw new JsonException("Properties other than $ref are not allowed in reference objects.");
```

Comment:
Does Json.NET require $ref as the one and only first property? Also it seems like other metadata that we don't recognize could come first. In general, we should align to Json.NET behavior, and also have tests that actually interop with Json.NET syntax (or use the same input data) -- we do this quite a bit in the reader\writer tests and a little in the serializer custom converter tests.

Reply:
Json.Net allows other properties to be before $ref but I dont think we shall allow that. However you are right, I still need to validate that other known metadata properties ($id) are not found before $ref.